Improve HTTP-Cache doc (#1448)
ezsystems/developer-documentation: docs/guide/cache/content_aware_cache.md

    Tags are not prefixed for default repository (index "0").

The content tags are returned in a header in the responses from [[= product_name =]]. The header name is dependent on
which HTTP Cache [[= product_name =]] is configured with:

> - Symfony Reverse proxy : `X-Cache-Tags`

Author: Fixed in https://github.com/ezsystems/developer-documentation/pull/1448/commits/415d7cda69ef7c931037d9675258efc9febc7480

> - Symfony Reverse proxy : `X-Cache-Tags`
> - Varnish : `xkey`
> - Fastly : `Surrogate-Key`

Reviewer: ```suggestion
- Symfony reverse proxy: `X-Cache-Tags`
- Varnish: `xkey`
- Fastly: `Surrogate-Key`
```

`ezplatform-http-cache` uses Repository API event subscribers to listen to events emitted on Repository operations,
and depending on the operation triggers expiry on a specific tag or set of tags.
All event subscribers can be found in `ezplatform-http-cache/src/EventSubscriber/CachePurge`.

### Understanding when the different tags are purged

#### Tags that are purged during a publish event.

> Below is an example of a Content structure. The tags that the content-view controller will tag each location with is
> also listed

Reviewer: ```suggestion
Below is an example of a Content structure. The tags which the content view controller adds to each location are
also listed
```
a bit simpler

```
   - [Home] (content-id=52, location-id=2)
     ez-all c52 ct42 l2 pl1 p1 p2
     |
     - [Parent1](content-id=53, location-id=20)
       ez-all c53 ct1 l20 pl2 p1 p2 p20
       |
       - [Child](content-id=55, location-id=22)
         ez-all c55 ct1 l22 pl20 p1 p2 p20 p22
     - [Parent2](content-id=54, location-id=21)
       ez-all c55 ct1 l22 pl2 p1 p2 p22
```

In the event that a new version of `Child` is published, the following keys will be purged:
- `c55`, because Content `[Child]` was changed
- `r55`, because cache for any object that has a relation to Content `[Child]` should be purged
- `l22`, because Location `[Child]` has changed ( that would be location holding content-id=55)
- `pl22`, because cache for children of `[Child]` should be purged
- `rl22`, because cache for any object that has a relation to Location `[Child]` should be purged
- `l20`, because cache for parent of `[Child]` should be purged

> - `pl20`, because cache for siblings of `[Child]` should be purged

Reviewer: I don't think we need whitelines if it's a list, markdown should render it correctly anyway

> - `pl20`, because cache for siblings of `[Child]` should be purged

Author: FYI : I am leaving all suggestions containing bullets with whitelines as-is, until you decide if those whitelines should be there, or until they are removed from the suggestions

> - `pl20`, because cache for siblings of `[Child]` should be purged

Reviewer 2: So in this case please ignore my suggestion.

> - `pl20`, because cache for siblings of `[Child]` should be purged

Author: okay, marking this discussion as solved then

> In words, HTTP Cache for any location representing `[Child]`, any Content that relates to the Content `[Child]`, the 

Reviewer: ```suggestion
In other words, HTTP Cache for any location representing `[Child]`, any Content that relates to the Content `[Child]`, the 
```
or
```suggestion
In summary, HTTP Cache for any location representing `[Child]`, any Content that relates to the Content `[Child]`, the 
```

`[Parent1]`, any children on `[Parent1]`.
Effectively, in this example HTTP cache for `[Parent1]` and `[Child]` will be cleared.

#### Tags that are purged during a move event.

With the same Content structure as above, the `[Child]` location is moved below `[Parent2]`.

The new structure will then be:
```
   - [Home] (content-id=52, location-id=2)
     ez-all c52 ct42 l2 pl1 p1 p2
     |
     - [Parent1](content-id=53, location-id=20)
       ez-all c53 ct1 l20 pl2 p1 p2 p20
     - [Parent2](content-id=54, location-id=21)
       ez-all c55 ct1 l22 pl2 p1 p2 p22
       |
       - [Child](content-id=55, location-id=22)
         ez-all c55 ct1 l22 pl21 p1 p2 p21 p22
```

The following keys will be purged during the move:
- `l20`, because cache for previous parent of `[Child]` should be purged (`[Parent1]`)
- `pl20`, because cache for children of `[Parent1]` should be purged
- `l21`, because cache for new parent of `[Child]` should be purged (`[Parent2]`)
- `pl21`, because cache for all children of new parent (`[Parent2]`) should be purged
- `p22`, because cache for any element below `[Child]` should be purged (because path has changed)

> 

Reviewer: as above

> 

Author: okay, marking this discussion as solved then

> In Word, HTTP Cache for `[Parent1]`, children of `[Parent1]` ( if any ), `[Parent2]`, children of `[Parent2]` ( if any ),

Reviewer: ```suggestion
In other words, HTTP Cache for `[Parent1]`, children of `[Parent1]` ( if any ), `[Parent2]`, children of `[Parent2]` ( if any ),
```

that .vcl code for Varnish V6.x will likely not work as-is on Fastly.

In this chapter we'll look into how to debug problems related to HTTP-cache. In order to that we must be able to look at
the responses and headers [[= product_name =]] sends to the HTTP Cache, and not so much on what responses and headers
the HTTP Cache sends the client (web-browser).
That means that you must be able to send requests to your origin (webserver), requests that does *not* go through Varnish/Fastly.
If you run nginx and Varnish on premise, you likely know what host and port number both Varnish and nginx runs on. If you
are testing on a Fastly enabled environment on Ibexa Cloud provided by Platform.sh, you'll need to use the Platform.sh
Dashboard in order to obtain the endpoint for nginx.

In the example below we'll debug and figure out why a given frontpage is not cached properly by Fastly. We have learned
that by running the following command *multiple* times:

```curl -IXGET https://www.staging.foobar.com.us-2.platformsh.site```

And it always outputs

```
HTTP/2 200
(...)
x-cache: MISS
```

### Finding the nginx endpoint on Platform.sh

> Assuming you on Platform have a project with id=`asd123asd123` and an environment named `Staging` which you access via Fastly on

Author: @Oc7at : Do you have some input for a straight forward process for how to get the route to nginx endpoing on a fastly-enabled environment? I found it to be slightly different depending on if you are on grid vs dedicated etc ( on dedicated, i think I in the past found the region host (......eu2.platformsh.site or something), and then just added `gw.` in front...

Reviewer 3: If you're looking for the origin (or "backend" as it's the term used by Fastly), you'll indeed want `c.<clusterid>.ent.platform.sh` for dedicated clusters. For non-dedicated projects, you can use [our documentation page](https://docs.platform.sh/development/public-ips.html) that lists all the IPs of our regions; you can usually use `gw.<region>.platformsh.site` though, like `gw.eu-2.platformsh.site` as you mention. Another good way of doing that for non-dedicated environments with our CLI is to use `platform environment:info edge_hostname`, as stated [on this page](https://docs.platform.sh/domains/steps/dns.html?#where-should-the-cname-point-to).

    Similarly to purging from code, the tags you purge on, are prefixed to match the currently configured SiteAccess. 
    When you use this command in combination with multi-repository setup, make sure to specify SiteAccess argument.

## Testing and Debugging HTTP cache

> It is important to test your code in an environment which is as similar as your production environment as possible. That
> means that if only are testing locally using the default Symfony Reverse proxy when your are going to use Varnish or
> Fastly in production, you are likely edning up some (bad) surprises. Due to the symfony reverse proxy's lack of support for ESIs, it behaves
> quite different from Varnish and Fastly in some aspects.
> If you are going to use Varnish in production, make sure you also test your code with Varnish.
> If you are going to use Fastly in production, testing with Fastly in your developer install is likely not feasible
> (you're local development environment must then be accessible for Fastly). Testing with Varnish instead will in most
> cases do the job. But if you need to change the varnish configuration to make your site work, be aware that Varnish and Fastly uses different dialects, and
> that .vcl code for Varnish V6.x will likely not work as-is on Fastly.

Reviewer: ```suggestion
You should test your code in an environment similar to your production environment.
	If you only test locally using the default Symfony reverse proxy when you want to use Varnish or
	Fastly in production, you may encounter problems. Symfony reverse proxy does not support ESI, it behaves differently than Varnish and Fastly.
	If you want to use Varnish in production, make sure you also test your code with Varnish.
	If you want to use Fastly in production, testing with Fastly in your developer installation is likely not feasible
	(local development environment must be then accessible for Fastly). Testing with Varnish instead works in most	cases.
	But if you need to change the Varnish configuration to make your site work, be aware that Varnish and Fastly use different dialects, and
	that .vcl code for Varnish 6.x may not work as it does in Fastly.
```

> that .vcl code for Varnish V6.x will likely not work as-is on Fastly.

Author: In the last sentence here.. now you say it may not work in varnish in the same way "as it does in Fastly", right? It is the opposite of my original intention

> that .vcl code for Varnish V6.x will likely not work as-is on Fastly.

Reviewer 2: I corrected the last sentence.

> In this chapter we'll look into how to debug problems related to HTTP-cache. In order to that we must be able to look at
> the responses and headers [[= product_name =]] sends to the HTTP Cache, and not so much on what responses and headers
> the HTTP Cache sends the client (web-browser).
> That means that you must be able to send requests to your origin (webserver), requests that does *not* go through Varnish/Fastly.
> If you run nginx and Varnish on premise, you likely know what host and port number both Varnish and nginx runs on. If you
> are testing on a Fastly enabled environment on Ibexa Cloud provided by Platform.sh, you'll need to use the Platform.sh
> Dashboard in order to obtain the endpoint for nginx.

Reviewer: ```suggestion
This section describes to how to debug problems related to HTTP cache. 
	In order to that, you must be able to look both at
	responses and headers [[= product_name =]] sends to HTTP cache, and not so much at responses and headers
	the HTTP cache sends to the client (web browser).
	It means you must be able to send requests to your origin (web server) that do not go through Varnish or Fastly.
	If you run Nginx and Varnish on premise, you should know what host and port number both Varnish and Nginx runs on. If you
	perform tests on Fastly enabled environment on Ibexa Cloud provided by Platform.sh, you need to use the Platform.sh
	Dashboard to obtain the endpoint for Nginx.
```

> In the example below we'll debug and figure out why a given frontpage is not cached properly by Fastly. We have learned
> that by running the following command *multiple* times:
> 
> ```curl -IXGET https://www.staging.foobar.com.us-2.platformsh.site```
> 
> And it always outputs

Reviewer: ```suggestion
The following example shows how to debug and check why Fastly does not cache the front page properly. 
If you run the command multiple times:

`curl -IXGET https://www.staging.foobar.com.us-2.platformsh.site`

it always outputs:
```

```
HTTP/2 200
(...)
x-cache: MISS

> ```

Author: This doesn't look right to me:
> If you run the command multiple times:
that by running the following command **multiple** times:

> ```

Reviewer: You are right, something got mixed up, but here's the proposed change: https://github.com/ezsystems/developer-documentation/pull/1448/files#r744586659

Typically, you may just add a `gw` to the hostname and use nslookup to look it up

```bash
    $ nslookup
    > gw.us-2.platformsh.site
   (...)
   Address:  1.2.3.4
```

You may also use the [Platform.sh CLI command](https://docs.platform.sh/development/cli.html) to find [the endpoint](https://docs.platform.sh/domains/steps/dns.html?#where-should-the-cname-point-to) :

```bash
    $ platform environment:info edge_hostname
```

#### Finding nginx endpoint on dedicated cloud

If you have a dedicated 3-node cluster on Platform.sh, the procedure for getting the endpoint to environments that are 
located on that cluster (`production` and sometimes also `staging`) is slightly different:
In the `URLs` dropdown in the Platform.sh dashboard you need to find the route that has the format 
`something.[clusterid].ent.platform.sh/`, for instance `myenvironment.asddfs2323.ent.platform.sh/`

The endpoint will in that case be in the format `c.[clusterid].ent.platform.sh`, for instance `c.asddfs2323.ent.platform.sh/`
Next, use nslookup to find the IP:

```bash
    $ nslookup
    > c.asddfs2323.ent.platform.sh
   (...)
   Address:  1.2.3.4
```

### Fetching the User-Context-Hash

> As explained in [User Context Hash caching](#user-context-hash-caching), the HTTP Cache indexes the cache based on the
> user-context-hash. Users with the same user-context-hash will share the same cache (as long as [[= product_name =]]
> responds with `Vary: X-User-Hash`).
> 
> So in order to simulate the requests the HTTP-Cache sends to [[= product_name =]], we need that user-context-hash.
> That can be obtained using `curl`

Reviewer 2: ```suggestion
As explained in [User Context Hash caching](guide/cache/context_aware_cache.md#user-context-hash-caching), the HTTP cache indexes the cache based on the
user-context-hash. Users with the same user-context-hash here the same cache (as long as [[= product_name =]]
responds with `Vary: X-User-Hash`).

In order to simulate the requests the HTTP cache sends to [[= product_name =]], you need this user-context-hash.
To obtain it, use `curl`.
```

```bash
    $ curl -IXGET --resolve www.staging.foobar.com.us-2.platformsh.site:443:1.2.3.4 --header "Surrogate-Capability: abc=ESI/1.0" --header "accept: application/vnd.fos.user-context-hash" --header "x-fos-original-url: /" https://www.staging.foobar.com.us-2.platformsh.site/_fos_user_context_hash
```

> Some notes about each of these parameters:
> - `-IXGET`, one of many ways to tell curl that we want to send a GET request, but we are only interested in outputting the headers
> - `--resolve www.staging.foobar.com.us-2.platformsh.site:443:1.2.3.4`
>   - We tell curl not to do a DNS lookup for `www.staging.foobar.com.us-2.platformsh.site`. We do that because in our case
>     that will resolve to the Fastly endpoint, not our origin (nginx)
>   - We specify `443` because we are using `https`
>   - We provide the IP of the nginx endpoint at platform.sh (`1.2.3.4` in this example)
> - `--header "Surrogate-Capability: abc=ESI/1.0"`, strictly speaking not needed when fetching the user-context-hash, but this tells [[= product_name =]] that client understands ESI tags.
>   It is good practice to always include this header when imitating the HTTP Cache.
> - `--header "accept: application/vnd.fos.user-context-hash"` tells [[= product_name =]] that the client wants to receive the user-context-hash
> - `--header "x-fos-original-url: /"` is required by the fos-http-cache bundle in order to deliver the user-context-hash
> - `https://your-page-blah-blah.us-2.platformsh.site/_fos_user_context_hash` : here we use the hostname we earlier told
>   curl how to resolve using `---resolve`. `/_fos_user_context_hash` is the route to the controller that are able to
>   deliver the user-context-hash.
> - You may also provide the session cookie (`--cookie ".....=....") for a logged-in-user if you are interested in
>   the x-user-context-hash for a different user but anonymous
> 

Reviewer 2: ```suggestion
Parameters description:

- `-IXGET`, one methods to tell curl that you want to send a GET request, but you are interested in outputting the headers

- `--resolve www.staging.foobar.com.us-2.platformsh.site:443:1.2.3.4`
  - you tell curl not to do a DNS lookup for `www.staging.foobar.com.us-2.platformsh.site`. It is done because in this case
    it resolves to the Fastly endpoint, not the origin (Nginx)
  - You specify `443` because you use `https`
  - You provide the IP of the Nginx endpoint at Platform.sh (`1.2.3.4` in this example)
  
- `--header "Surrogate-Capability: abc=ESI/1.0"`, not needed when fetching the user-context-hash, but it tells [[= product_name =]] that client understands ESI tags.
  It is good practice to always include this header when imitating the HTTP cache.
  
- `--header "accept: application/vnd.fos.user-context-hash"` tells [[= product_name =]] that the client wants to receive the user-context-hash

- `--header "x-fos-original-url: /"` is required by the fos-http-cache bundle in order to deliver the user-context-hash

- `https://your-page-blah-blah.us-2.platformsh.site/_fos_user_context_hash` - it uses the hostname that was earlier told
  curl how to resolve using `---resolve`. `/_fos_user_context_hash` is the route to the controller that is able to
  deliver the user-context-hash.
  
- You can also provide the session cookie (`--cookie ".....=....") for a logged-in-user if you are interested in
  the x-user-context-hash for a different user but anonymous.

```

> The output for the given command should look similar to this:
> ```
>     HTTP/1.1 200 OK
>     Server: nginx/1.20.0
>     Content-Type: application/vnd.fos.user-context-hash
>     Transfer-Encoding: chunked
>     Connection: keep-alive
>     X-User-Hash: daea248406c0043e62997b37292bf93a8c91434e8661484983408897acd93814
>     Cache-Control: max-age=600, public
>     Date: Tue, 31 Aug 2021 13:35:00 GMT
>     Vary: Origin
>     Vary: cookie
>     Vary: authorization
>     X-Cache-Debug: 1
>     Surrogate-Key: ez-user-context-hash ez-all fos_http_cache_hashlookup-
> ```

Reviewer 2: ```suggestion
The output for this command should look similar to this:

```
    HTTP/1.1 200 OK
    Server: nginx/1.20.0
    Content-Type: application/vnd.fos.user-context-hash
    Transfer-Encoding: chunked
    Connection: keep-alive
    X-User-Hash: daea248406c0043e62997b37292bf93a8c91434e8661484983408897acd93814
    Cache-Control: max-age=600, public
    Date: Tue, 31 Aug 2021 13:35:00 GMT
    Vary: Origin
    Vary: cookie
    Vary: authorization
    X-Cache-Debug: 1
    Surrogate-Key: ez-user-context-hash ez-all fos_http_cache_hashlookup-
```

> The header `X-User-Hash` is the one we are interested in here, but you may also note the `Surrogate-Key` which
> holds the [cache tags](#understanding-cache-tags).

Reviewer 2: ```suggestion
The header `X-User-Hash` is the one of the interest here, but you may also note the `Surrogate-Key` which
holds the [cache tags](#understanding-cache-tags).
```

> ### Fetching the HTML response

Reviewer 2: ```suggestion
### Fetch the HTML response
```

> Now that we have the user-context-hash, we can ask origin for the actual resource we are after:
> 
> ```bash
>     $ curl -IXGET --resolve www.staging.foobar.com.us-2.platformsh.site:443:1.2.3.4 --header "Surrogate-Capability: abc=ESI/1.0" --header "x-user-hash: daea248406c0043e62997b37292bf93a8c91434e8661484983408897acd93814" https://www.staging.foobar.com.us-2.platformsh.site/
> ```
> 
> The output :

Reviewer 2: ```suggestion
Nowyou have the user-context-hash, and you can ask origin for the actual resource you are after:

```bash
    $ curl -IXGET --resolve www.staging.foobar.com.us-2.platformsh.site:443:1.2.3.4 --header "Surrogate-Capability: abc=ESI/1.0" --header "x-user-hash: daea248406c0043e62997b37292bf93a8c91434e8661484983408897acd93814" https://www.staging.foobar.com.us-2.platformsh.site/
```

The output :

```

Reviewer 2: @vidarl by word: actual you mean the current one or the real one?

Author: I don't understand your question.

But I guess "Nowyou" ---> "Now you" ?

Reviewer 2: @vidarl In the question I mean: does the word 'actual' stands for the "existing in fact or reality" or for "current, up to date"

```
HTTP/1.1 200 OK
Server: nginx/1.20.0
Content-Type: text/html; charset=UTF-8
Transfer-Encoding: chunked
Connection: keep-alive
Cache-Control: public, s-maxage=86400
Date: Wed, 01 Sep 2021 07:18:27 GMT
X-Cache-Debug: 1
Vary: X-User-Hash
Vary: X-Editorial-Mode
Surrogate-Control: content="ESI/1.0"
Surrogate-Key: ez-all c52 ct42 l2 pl1 p1 p2 r56 r57
```

> The `Cache-Control` header tells the HTTP cache to store the result in the cache for 1 day (86400 seconds)
> The `Vary: X-User-Hash` header tells the HTTP cache that this cache element may be used for all users which has
> the given `x-user-hash` (`daea248406c0043e62997b37292bf93a8c91434e8661484983408897acd93814`).
> The document might also be removed from the cache by purging any of the keys provided in the `Surrogate-Key` header.

Reviewer 2: ```suggestion

- `Cache-Control` header tells the HTTP cache to store the result in the cache for 1 day (86400 seconds)

- `Vary: X-User-Hash` header tells the HTTP cache that this cache element may be used for all users which have
the given `x-user-hash` (`daea248406c0043e62997b37292bf93a8c91434e8661484983408897acd93814`).

The document can also be removed from the cache by purging any of the keys provided in the `Surrogate-Key` header.
```

> So back to the original problem here. This resource is for some reason not cached by Fastly ( remember the
> `x-cache: MISS` we started with). But origin says this page can be cached for 1 day. How can that be?
> The likely reason is that this page also contains some ESI fragments and that one or more of these are not cachable.
> 
> So, first let's see if there are any ESIs here. We remove the `-IXGET` options (in order to see content of the response,
> not only headers) to curl and search for esi:

Reviewer 2: ```suggestion
Coming back to the original problem here. This resource is for some reason not cached by Fastly (see the
`x-cache: MISS` you started with). However origin says this page can be cached for 1 day. 
The reason for this is that this page also contains some ESI fragments and that one or more of these are not cachable.

First, check for any ESIs here. Remove the `-IXGET` options (in order to see content of the response,
not only headers) to curl and search for ESI:
```

Author: You removed "likely" from "likely reason". At this time in the debugging session, this is the hypothesis we are going to validate next.

Reviewer 2: I put the word likely back.

```bash
    $ curl --resolve www.staging.foobar.com.us-2.platformsh.site:443:1.2.3.4 --header "Surrogate-Capability: abc=ESI/1.0" --header "x-user-hash: daea248406c0043e62997b37292bf93a8c91434e8661484983408897acd93814" https://www.staging.foobar.com.us-2.platformsh.site/ | grep esi
```

> Output is :

Reviewer 2: ```suggestion
The output is :
```

    <esi:include src="/_fragment?_hash=egcfVhka%2Beo80g%2B6ztYi12ebiaWfoWwKIACmSWVRqjI%3D&_path=locationId%3D2%26contentId%3D52%26blockId%3D12%26versionNo%3D3%26languageCode%3Deng-GB%26serialized_siteaccess%3D%257B%2522name%2522%253A%2522site%2522%252C%2522matchingType%2522%253A%2522default%2522%252C%2522matcher%2522%253Anull%252C%2522provider%2522%253Anull%257D%26serialized_siteaccess_matcher%3Dnull%26_format%3Dhtml%26_locale%3Den_GB%26_controller%3DEzSystems%255CEzPlatformPageFieldTypeBundle%255CController%255CBlockController%253A%253ArenderAction" />
    <esi:include src="/_fragment?_hash=lnKTnmv6bb1XpaMPWRjV3sNazbn9rDXskhjGae1BDw8%3D&_path=locationId%3D2%26contentId%3D52%26blockId%3D13%26versionNo%3D3%26languageCode%3Deng-GB%26serialized_siteaccess%3D%257B%2522name%2522%253A%2522site%2522%252C%2522matchingType%2522%253A%2522default%2522%252C%2522matcher%2522%253Anull%252C%2522provider%2522%253Anull%257D%26serialized_siteaccess_matcher%3Dnull%26_format%3Dhtml%26_locale%3Den_GB%26_controller%3DEzSystems%255CCustomBundle%255CController%255CFooController%253A%253AcustomAction" />
```

> Now we must investigate the response of each of these ESI fragments to understand what is going on. It is important to
> now put that URL in single quotes since the URLS to the ESIs include special characters that can be interpreted by the
> shell.

Reviewer 2: ```suggestion
Now, nvestigate the response of each of these ESI fragments to understand what is going on. It is important to
put that URL in single quotes as the URLS to the ESIs include special characters that can be interpreted by the
shell.
```

Reviewer: ```suggestion
Now, investigate the response of each of these ESI fragments to understand what is going on. It is important to
put that URL in single quotes as the URLS to the ESIs include special characters that can be interpreted by the
shell.
```

#### 1st ESI

```bash
    $ curl -IXGET --resolve www.staging.foobar.com.us-2.platformsh.site:443:1.2.3.4 --header "Surrogate-Capability: abc=ESI/1.0" --header "x-user-hash: daea248406c0043e62997b37292bf93a8c91434e8661484983408897acd93814" 'https://www.staging.foobar.com.us-2.platformsh.site/_fragment?_hash=B%2BLUWB2kxTCc6nc5aEEn0eEqBSFar%2Br6jNm8fvSKdWU%3D&_path=locationId%3D2%26contentId%3D52%26blockId%3D11%26versionNo%3D3%26languageCode%3Deng-GB%26serialized_siteaccess%3D%257B%2522name%2522%253A%2522site%2522%252C%2522matchingType%2522%253A%2522default%2522%252C%2522matcher%2522%253Anull%252C%2522provider%2522%253Anull%257D%26serialized_siteaccess_matcher%3Dnull%26_format%3Dhtml%26_locale%3Den_GB%26_controller%3DEzSystems%255CEzPlatformPageFieldTypeBundle%255CController%255CBlockController%253A%253ArenderAction'
```

> We can also note that this esi is handled by a controller in the `EzPlatformPageFieldTypeBundle` bundle provided by [[= product_name =]]

Reviewer 2: ```suggestion
You can also note that this ESI is handled by a controller in the `EzPlatformPageFieldTypeBundle` bundle provided by [[= product_name =]].
```

> Output:

Reviewer 2: ```suggestion
The output is:
```

```
HTTP/1.1 200 OK
Server: nginx/1.20.0
Content-Type: text/html; charset=UTF-8
Transfer-Encoding: chunked
Connection: keep-alive
Cache-Control: public, s-maxage=86400
Date: Wed, 01 Sep 2021 07:51:40 GMT
Vary: Origin
Vary: X-User-Hash
Vary: X-Editorial-Mode
X-Cache-Debug: 1
Surrogate-Key: ez-all c52 l2
```

> The headers here looks Okay and nothing indicates that this ESI won't be cached by the HTTP-Cache
> The second ESI gives a similar response so we'll jump right down to the 3rd ESI
> 

Reviewer 2: ```suggestion
The headers here look correct and do not indicate that this ESI will not be cached by the HTTP cache
The second ESI has a similar response.

```

```

This ESI is handled by a custom `FooController::customAction` and the output of the command is:

Output:

```
HTTP/1.1 200 OK
Server: nginx/1.20.0
Content-Type: text/html; charset=UTF-8
Transfer-Encoding: chunked
Connection: keep-alive
Set-Cookie: eZSESSID21232f297a57a5a743894a0e4a801fc3=asrpqgmh5ll5ssseca3cov8er7; path=/; HttpOnly; SameSite=lax
Cache-Control: public, s-maxage=86400
Date: Wed, 01 Sep 2021 07:51:40 GMT
Vary: Origin
Vary: X-User-Hash
Vary: X-Editorial-Mode
X-Cache-Debug: 1
Surrogate-Key: ez-all
```

> As you can see, the `Cache-Control` and `Vary` headers looks all good. The request is as mentioned handled by a custom
> controller and the `Surrogate-Key` only contains the default `ez-all` value. That is not a problem as long as the controller
> doesn't return values from any Content in the [[= product_name =]] Repository. If it does, the controller should also add
> the corresponding IDs to such objects in that header.

Reviewer 2: ```suggestion
The `Cache-Control` and `Vary` headers look correct. The request is handled by a custom controller and the `Surrogate-Key` only contains the default `ez-all` value. 
This is not a problem as long as the controller
does not return values from any Content in the [[= product_name =]] Repository. If it does, the controller should also add
the corresponding IDs to such objects in that header.
```

> The big problem here is however the `Set-Cookie`. A ESI fragment should never set a cookie because:
> - Clients will only receive the headers set in the "mother" document (the headers in the "/" response in this case)
> - Only the content of ESIs responses will be returned to the client. *No headers set in the ESI response will ever reach
>   the client*. ESI headers are only seen by the HTTP Cache
> - Symfony reverse proxy doesn't support ESIs at all, and any ESI calls (`render_esi()`) will implicitly be replaced by
>   sub-requests (`render()`). So any `Set-Cookie` *will* be sent to the client when using Symgony reverse proxy.
> - Fastly will flag that resource as "not cachable" since it did set a cookie at least once. Even though that endpoint
>   stops setting cookies, Fastly will still not cache that fragment. Any document referencing that ESI will be a `MISS`.
>   Fastly cache needs to be purged (`Purge-all` request) in order to remove this flag
> - This means that it is a bad idea to always initiate a session when loading the front-page

Reviewer 2: ```suggestion
The `Set-Cookie`here may cause the problem. A ESI fragment should never set a cookie because:
- Clients will only receive the headers set in the "mother" document (the headers in the "/" response in this case).

- Only the content of ESIs responses will be returned to the client. **No headers set in the ESI response will ever reach the client**. ESI headers are only seen by the HTTP cache.
  
- Symfony reverse proxy does not support ESIs at all, and any ESI calls (`render_esi()`) will implicitly be replaced by
  sub-requests (`render()`). So any `Set-Cookie` **will** be sent to the client when using Symfony reverse proxy.
  
- Fastly will flag it resource as "not cachable" because it set a cookie at least once. Even though that endpoint.
  stops setting cookies, Fastly will still not cache that fragment. Any document referring to that ESI will be a `MISS`.
  Fastly cache needs to be purged (`Purge-all` request) in order to remove this flag.
  
- It means that it is not recommended to always initiate a session when loading the front page.
```

Author: FYI :Edited your suggestion : "the `Set-Cookie`here" --> "the `Set-Cookie` here"

> Therefore, you must ensure yourself that you do not unintendedly starts a session in a controller used by ESIs by for
> instance trying to access as session variable before a session has been initiated yet.

Reviewer 2: ```suggestion
You must ensure that you do not unintendedly start a session in a controller used by ESIs, for example, when trying to access as session variable before a session has been initiated yet.
```

    that will resolve to the Fastly endpoint, not our origin (nginx)
  - We specify `443` because we are using `https`
  - We provide the IP of the nginx endpoint at platform.sh (`1.2.3.4` in this example)
- `--header "Surrogate-Capability: abc=ESI/1.0"`, strictly speaking not needed when fetching the user-context-hash, but this tells [[= product_name =]] that client understands ESI tags.
  It is good practice to always include this header when imitating the HTTP Cache.
- `--header "accept: application/vnd.fos.user-context-hash"` tells [[= product_name =]] that the client wants to receive the user-context-hash
- `--header "x-fos-original-url: /"` is required by the fos-http-cache bundle in order to deliver the user-context-hash
- `https://your-page-blah-blah.us-2.platformsh.site/_fos_user_context_hash` : here we use the hostname we earlier told
  curl how to resolve using `---resolve`. `/_fos_user_context_hash` is the route to the controller that are able to
  deliver the user-context-hash.
- You may also provide the session cookie (`--cookie ".....=....") for a logged-in-user if you are interested in
  the x-user-context-hash for a different user but anonymous

> The output for the given command should look similar to this:

Author: ```suggestion
The output for this command should look similar to this:
```

- `https://your-page-blah-blah.us-2.platformsh.site/_fos_user_context_hash` : here we use the hostname we earlier told
  curl how to resolve using `---resolve`. `/_fos_user_context_hash` is the route to the controller that are able to
  deliver the user-context-hash.
- You may also provide the session cookie (`--cookie ".....=....") for a logged-in-user if you are interested in
  the x-user-context-hash for a different user but anonymous

The output for this command should look similar to this:
```
    HTTP/1.1 200 OK
    Server: nginx/1.20.0
    Content-Type: application/vnd.fos.user-context-hash
    Transfer-Encoding: chunked
    Connection: keep-alive
    X-User-Hash: daea248406c0043e62997b37292bf93a8c91434e8661484983408897acd93814
    Cache-Control: max-age=600, public
    Date: Tue, 31 Aug 2021 13:35:00 GMT
    Vary: Origin
    Vary: cookie
    Vary: authorization
    X-Cache-Debug: 1
    Surrogate-Key: ez-user-context-hash ez-all fos_http_cache_hashlookup-
```

The header `X-User-Hash` is the one of the interest here, but you may also note the `Surrogate-Key` which
holds the [cache tags](#understanding-cache-tags).

### Fetch the HTML response

> Now that we have the user-context-hash, we can ask origin for the actual resource we are after:

Reviewer 2: ```suggestion
Now you have the user-context-hash, and you can ask origin for the actual resource you are after:
```
